### PR TITLE
Llvm14config

### PR DESCRIPTION
**Pull request by nishant-sachdeva**

### CODE REVIEW

File: src/utils.cpp [hunk 1]
```diff
@@ -82,6 +82,19 @@ std::string IR2Vec::getDemagledName(const llvm::Function *function) {
   return demangledName;
 }
 
+// Function to get demanged instruction name
+std::string IR2Vec::getDemagledName(const llvm::Instruction *instruction) {
```
Comment (svkeerthy) [inline src/utils.cpp:86]: Why is this method needed?

File: seed_embeddings/OpenKE/generate_embedding_ray.py
Comment (svkeerthy) [inline seed_embeddings/OpenKE/generate_embedding_ray.py:1]: Is this file needed in this version?

**Review by svkeerthy — CHANGES_REQUESTED** [review]

File: seed_embeddings/OpenKE/generate_embedding_ray.py
Comment (nishant-sachdeva) [reply seed_embeddings/OpenKE/generate_embedding_ray.py:1]: Yea. We've retrained the llvm-14 version with Ray as well.

File: src/utils.cpp (hunk 1 above)
Comment (nishant-sachdeva) [reply src/utils.cpp:86]: The `refactoredCore.cpp` has the method, `createInstructionVectorDict` . The method had duplicated code to get the `Demangled` Name for the instruction. So, in interest of better modularity, and reusability, I added this function here.

File: src/include/utils.h [hunk 2]
```diff
@@ -50,6 +50,7 @@ void collectDataFromVocab(std::map<std::string, Vector> &opcMap);
 void scaleVector(Vector &vec, float factor);
 // newly added
 std::string getDemagledName(const llvm::Function *function);
+std::string getDemagledName(const llvm::Instruction *instruction);
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply src/include/utils.h:53]: See https://github.com/IITH-Compilers/IR2Vec/pull/80#discussion_r1435667043

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/refactoredCore.cpp
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/refactoredCore.cpp:1]: Done in the latest push

File: .gitignore [hunk 3]
```diff
@@ -4,5 +4,6 @@ __pycache__/
 .vscode
 experiments/*/output
 seed_embeddings/triplets.txt
+seed_embeddings/triplets14.txt
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply .gitignore:7]: Indeed, not required. Removed in the latest push

File: Manylinux2014_Compliant_Source/pkg/build.sh [hunk 4]
```diff
@@ -18,8 +18,8 @@ cd ..
 cmake -DCMAKE_BUILD_TYPE=Release ../src && make -j"$(nproc)" && make install
 
 cd ..
-cp src/include/utils.h Manylinux2014_Compliant_Source/pkg/IR2Vec/
+cp src/include/*.h Manylinux2014_Compliant_Source/pkg/IR2Vec/
 cp build/version.h Manylinux2014_Compliant_Source/pkg/IR2Vec/
-cp vocabulary/seedEmbeddingVocab-*.txt Manylinux2014_Compliant_Source/pkg/IR2Vec/
+cp vocabulary/seedEmbeddingVocab-llvm14.txt Manylinux2014_Compliant_Source/pkg/IR2Vec/
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/build.sh:23]: IMO, this way of naming is better instead of inserting an asterisk. That method is prone to errors. 

In `testIR2Vec.py`, we have 
```
SEED_VERSION = (
    list(IR2Vec.preparation.install_loc.glob("seed*.txt"))[0]
    .name.split(".")[0]
    .split("-")[-1]
)
```

If in future, for any reason, we have multiple files with this naming scheme, the code becomes error prone.

[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/build.sh:23]: However, if it's about the specific naming of llvm14, we can change it to a different neutral name

File: src/include/IR2Vec.h [hunk 5]
```diff
@@ -29,17 +29,19 @@ class Embeddings {
...
-  // Use this constructor if the representations ought to be written to a file.
-  // Analogous to the command line options that are being used in IR2Vec binary.
+  // Use this constructor if the representations ought to be written to a
+  // file. Analogous to the command line options that are being used in IR2Vec
+  // binary.
   Embeddings(llvm::Module &M, IR2VecMode mode, std::string vocab, char level,
-             std::ostream *o, float WO = 1, float WA = 0.2, float WT = 0.5,
-             std::string funcName = "") {
+             std::ostream *o, std::string funcName = "", float WO = 1,
+             float WA = 0.2, float WT = 0.5) {
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply src/include/IR2Vec.h:44]: Imo it would be better to keep the order of the two polymorphed functions as similar as possible. Plus, in our use cases, we are passing string values to the funcName parameter, and are using default values for WO, WA, WT variables. So I thought it better to keep them in the last, as the other polymorphed function does

File: README.md [hunk 6]
```diff
@@ -47,28 +49,17 @@ Please see [here](https://compilers.cse.iith.ac.in/projects/ir2vec/) for more de
...
-`IR2Vec` can be installed in different ways to accommodate individual preferences and requirements effectively. You may select to install via a user-friendly Python wheel setup if you are a Python user, or opt for a C++ based installation if you are looking to integrate with a compiler pass or necessitate advanced control and enhanced integration capabilities. The detailed setup steps are mentioned in the following sections.
-
-## Python
-
-If you prefer working with Python, you can easily install `IR2Vec` using `pip`.
+Install the Latest release of the Python package using:
 
 ```
-pip install -U ir2vec
+pip install -U IR2Vec
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply README.md:55]: This seems to be correct, as per https://pypi.org/project/IR2Vec/

File: README.md [hunk 7]
```diff
@@ -47,28 +49,17 @@ Please see [here](https://compilers.cse.iith.ac.in/projects/ir2vec/) for more de
...
-## Cpp
-
-If you're a C++ developer and require low-level control, optimization, or integration with C++ projects, you can build `IR2Vec` from source. First, ensure the below requirements are satisfied, then follow the steps mentioned in the [Building from source](#building-from-source) section.
+However, building from source is also supported for which the requirements and steps are listed down below.
 
 ## Requirements
 * cmake (>= 3.13.4)
 * GNU Make (4.2.1)
-* LLVM (14.0.0) - [src](https://github.com/llvm/llvm-project/tree/release/14.x), [release](https://releases.llvm.org/download.html#14.0.0)
+* LLVM (14.0.1) - [src](https://github.com/llvm/llvm-project/tree/release/14.x), [release](https://releases.llvm.org/download.html#14.0.0)
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply README.md:62]: Done in latest push

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 8]
```diff
@@ -47,342 +51,366 @@
...
-      PyDict_SetItemString(Embedding_dict, "Status", Py_False);
-      return Embedding_dict;
-    } else {
-      if (level[0] != 'p' && level[0] != 'f') {
-        PyDict_SetItemString(
-            Embedding_dict, "Message",
-            PyUnicode_FromString("Invalid level specified: Use either p or f"));
-        PyDict_SetItemString(Embedding_dict, "Status", Py_False);
-        return Embedding_dict;
+      PyObject *temp3 = PyList_New(0);
```
Comment (svkeerthy) [inline Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:125]: Better name than temp3?

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 9]
```diff
@@ -47,342 +51,366 @@
...
-    size_t sz = 17;
-    int status = 0;
-    // coying llvm samll map vector data into c++ map
-
-    // for InstVecMap
-    for (auto Inst_it : llvm_inst_vec_map) {
+    for (auto &Inst_it : llvm_inst_vec_map) {
+      std::string demangledName = IR2Vec::getDemagledName(Inst_it.first);
+
+      PyObject *temp3 = PyList_New(0);
```
Comment (svkeerthy) [inline Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:158]: Same here please..

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 10]
```diff
@@ -47,342 +51,366 @@
...
+  // generateEncodings
+  PyObject *generateEncodings(OpType type, std::string funcName = "") {
+    // Invokinng IR2Vec lib exposed functions
+    IR2Vec::iname = this->fileName;
+    IR2Vec::IR2VecMode ir2vecMode =
+        (this->mode == string("sym") ? IR2Vec::Symbolic : IR2Vec::FlowAware);
+    // The scope of this Module object is extremely crucial
+    std::unique_ptr<llvm::Module> Module;
+    Module = IR2Vec::getLLVMIR();
+    std::string vocab_path = seed_emb_path + "/seedEmbeddingVocab-llvm14.txt";
```
Comment (svkeerthy) [inline Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:179]: Is it possible not to hardcode the file name at this level?

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 11]
```diff
@@ -47,342 +51,366 @@
...
+typedef struct {
+  PyObject_VAR_HEAD ir2vecHandler *ir2vecObj;
+} ir2vecHandlerObject;
 
-    // all c++ container are ready to convert
-    // iterate over a c++ container , if it has embedd anotehr c++c ocntainer
-    // then iterate over it also
+PyObject *getInstructionVectors(ir2vecHandlerObject *self, PyObject *args) {
+  // check for args, and null etc
+  if ((self->ir2vecObj) == NULL) {
```
Comment (svkeerthy) [inline Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:225]: Should these checks be on NULL or nullptr?

Comment (svkeerthy) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:225]: Please check. These checks are in multiple places.

File: Manylinux2014_Compliant_Source/pkg/build.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (svkeerthy) [reply Manylinux2014_Compliant_Source/pkg/build.sh:23]: Is it better to have it just as `seedEmbeddingVocab.txt`?

File: Manylinux2014_Compliant_Source/readme.md
Comment (svkeerthy) [inline Manylinux2014_Compliant_Source/readme.md:1]: This README is no more relevant. Please delete this file.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (svkeerthy) [reply README.md]: Ok. Both works.

File: src/include/IR2Vec.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (svkeerthy) [reply src/include/IR2Vec.h:44]: Great. Makes sense.

File: src/utils.cpp (hunk 1 above)
Comment (svkeerthy) [reply src/utils.cpp:86]: What does demangling instruction mean?

**Review by svkeerthy — CHANGES_REQUESTED** [review]

File: Manylinux2014_Compliant_Source/pkg/build.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/build.sh:23]: Yea, that can be done

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 12]
```diff
@@ -6,7 +6,10 @@
 //
 
 #define PY_SSIZE_T_CLEAN
+#include "CollectIR.h"
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:9]: Nope, these were included for some direct testing during the refactoring. Will remove

File: seed_embeddings/README.md [hunk 13]
```diff
@@ -15,7 +15,7 @@ If you have not done `make`, follow the following steps to build `ir2vec` binary
 ## Step 2: Generating Triplets
 #### Steps to collect the triplets
  Run `triplets.sh` script with the required parameters
-`bash triplets.sh <buildDir> <numOpt> <llFileList> <outputFileName>`
+ Usage: `bash triplets.sh <build dir> <No of opt> <llFile alist> <output FileNme>`
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply seed_embeddings/README.md:18]: Yea. Fixed

File: seed_embeddings/triplets.sh [hunk 14]
```diff
@@ -4,7 +4,7 @@
 # This software is available under the BSD 4-Clause License. Please see LICENSE
 # file in the top-level directory for more details.
 #
-#Usage: bash triplets.sh <build dir> <No of opt> <llFile list> <output FileName>
+#Usage: bash triplets.sh <build dir> <No of opt> <llFile alist> <output FileNme> <llvm_build>
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply seed_embeddings/triplets.sh:7]: Done

File: seed_embeddings/triplets.sh [hunk 15]
```diff
@@ -35,7 +35,7 @@ if [ -z $4 ]; then
 	exit
 fi
 
-LLVM_BUILD=$5
+LLVM_BUILD="/usr"
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply seed_embeddings/triplets.sh:38]: Done

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 16]
```diff
@@ -47,342 +51,366 @@
...
+      : fileName(fileName), outputFile(outputFile), mode(mode), level(level) {}
+
+  std::string getFile() { return fileName; }
+  std::string getOutputFile() { return outputFile; }
+  std::string getMode() { return mode; }
+  std::string getLevel() { return level; }
+
+  // Function to get Program Vector List
+  PyObject *
+  createProgramVectorList(llvm::SmallVector<double, DIM> llvm_pgm_vec) {
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:102]: Done

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp (hunk 8 above)
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:122]: Done

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp (hunk 9 above)
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:155]: Done

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp [hunk 17]
```diff
@@ -47,342 +51,366 @@
...
+  return false;
+}
 
-  const char *filename = "\0"; // need to generate module object from this .
-                               // Thus , need to add one more .cpp in .so
-  const char *mode = "\0";
+// create Enum with three options : Program, Function, Instruction
+enum class OpType { Program, Function, Instruction };
+
+class ir2vecHandler {
```
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp:83]: `ir2vecHandler` is already camelCasing isn't it ? Leaving this as it is for now

File: src/utils.cpp (hunk 1 above)
Comment (nishant-sachdeva) [reply src/utils.cpp:86]: The code was used to get the `demangled`  format name for the instruction. I shifted that code to the include files to improve modularity

File: Manylinux2014_Compliant_Source/readme.md
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/readme.md:1]: Done

File: seed_embeddings/OpenKE/analogy.py
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply seed_embeddings/OpenKE/analogy.py:1]: Done

File: src/include/utils.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply src/include/utils.h]: Removed in the latest push. Updated API output for `InstructionVectors` . Added tests for the same

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp]: Corrected.

File: Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp (hunk 11 above)
Comment (nishant-sachdeva) [reply Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp]: My bad. 

I've pushed the following changes here
1. I've replaced `NULL` with `nullptr` wherever pointer comparisons are taking place, or where I expect deleted pointers.
2. Replacing instance of NULL with Py_RETURN_NONE wherever we're returning a None Type object to the Python API

**Review by svkeerthy — CHANGES_REQUESTED** [review]

**Review by svkeerthy — APPROVED** [review]
